libs/contrib/Test/Golden.idr

```diff
@@ -0,0 +1,194 @@
+|||
+||| + `run.sh` a *shell* script that runs the test
+||| + `expected` the expected output of `run.sh`
+|||
+||| During testing, the test harness will generate an artefact named `output` and use `git diff` to examine the two files.
+||| This assumes that the expected has been checked into `git`.
+||| If `git diff` failes then the runner will simply present the expected and 'given' files.
+|||
+||| # Options
+|||
+||| The test harness has several options that must be set:
+|||
+||| + `idris2` the path of the executable we are testing.
+||| + `onlyNames` the list of tests to run relative to the generated executable.
+||| + `interactive` Should we update the expected file or not.
+|||
+||| # Usage
+|||
+||| When compiled to an executable the expected usage is:
+|||
+|||```sh
+|||runtests <path to executable> [--interactive] [--only [NAMES]]
+|||```
+|||
+||| assuming that the test runner is compiler to an executable named `runtests`.
+module Test.Golden
+
+import Data.Maybe
+import Data.List
+import Data.List1
+import Data.Strings
+
+import System
+import System.Directory
+import System.File
+import System.Info
+import System.Path
+
+-- [ Options ]
+
+||| Options for the test driver.
+public export
+record Options where
+  constructor MkOptions
+  ||| Name of the idris2 executable
+  idris2      : String
+  ||| Should we only run some specific cases?
+  onlyNames   : List String
+  ||| Should we run the test suite interactively?
+  interactive : Bool
+
+export
+usage : String
+usage = "Usage: runtests <path> [--interactive] [--only [NAMES]]"
+
+||| Process the command line options.
+export
+options : List String -> Maybe Options
+options args = case args of
+    (_ :: idris2 :: rest) => go rest (MkOptions idris2 [] False)
+    _ => Nothing
+
+  where
+
+    go : List String -> Options -> Maybe Options
+    go rest opts = case rest of
+      []                      => pure opts
+      ("--interactive" :: xs) => go xs (record { interactive = True } opts)
+      ("--only" :: xs)        => pure $ record { onlyNames = xs } opts
+      _ => Nothing
+
+-- [ Core ]
+
+fail : String -> IO ()
+fail err
+    = do putStrLn err
+         exitWith (ExitFailure 1)
+
+
+||| Normalise strings between different OS.
+|||
+||| on Windows, we just ignore backslashes and slashes when comparing,
+||| similarity up to that is good enough. Leave errors that depend
+||| on the confusion of slashes and backslashes to unix machines.
+normalize : String -> String
+normalize str =
+    if isWindows
+      then pack $ filter (\ch => ch /= '/' && ch /= '\\') (unpack str)
+      else str
+
+||| Run the specified Golden test with the supplied options.
+|||
+||| See the module documentation for more information.
+export
+runTest : Options -> String -> IO Bool
+runTest opts testPath
+    = do changeDir testPath
+         isSuccess <- runTest'
+         changeDir "../.."
```

Comment:
This hardcodes the fact that the target paths are 2 levels of nesting deep.
And the fact that the golden test does not change that fact. Should we
do `pwd <- currentDir` (from `base`'s `System.Directory`) before `changeDir`
and change back to the saved `pwd` instead?

Reply:
i've changed the test runner to take in a string rather than "../.." this allows the user to decide the behaviour rather than us.